### PR TITLE
 BUG: SeriesGroupBy.apply sets name attribute if result is DataFrame

### LINKED ISSUE #49907: BUG: SeriesGroupBy.apply sets name attribute if result is DataFrame
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
def f(piece):
    return DataFrame({"value": piece, "demeaned": piece - piece.mean()})

df = pd.DataFrame({'a': [1,2,3], 'b': [4, 5, 5]})
res = df.groupby('a')['b'].apply(f)
print(res.name)
```

### Issue Description

This prints `'b'`. But `.name` shouldn't be an attribute of `DataFrame`...

### Expected Behavior

```
AttributeError: 'DataFrame' object has no attribute 'name'
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 3b09765bc9b141349a2f73edd15aaf4f436928bd
python           : 3.8.15.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.102.1-microsoft-standard-WSL2
Version          : #1 SMP Wed Mar 2 00:30:59 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 2.0.0.dev0+759.g3b09765bc9
numpy            : 1.23.5
pytz             : 2022.6
dateutil         : 2.8.2
setuptools       : 65.5.1
pip              : 22.3.1
Cython           : 0.29.32
pytest           : 7.2.0
hypothesis       : 6.58.0
sphinx           : 4.5.0
blosc            : None
feather          : None
xlsxwriter       : 3.0.3
lxml.etree       : 4.9.1
html5lib         : 1.1
pymysql          : 1.0.2
psycopg2         : 2.9.3
jinja2           : 3.0.3
IPython          : 8.6.0
pandas_datareader: 0.10.0
bs4              : 4.11.1
bottleneck       : 1.3.5
brotli           : 
fastparquet      : 2022.11.0
fsspec           : 2021.11.0
gcsfs            : 2021.11.0
matplotlib       : 3.6.2
numba            : 0.56.4
numexpr          : 2.8.3
odfpy            : None
openpyxl         : 3.0.10
pandas_gbq       : 0.17.9
pyarrow          : 9.0.0
pyreadstat       : 1.2.0
pyxlsb           : 1.0.10
s3fs             : 2021.11.0
scipy            : 1.9.3
snappy           : 
sqlalchemy       : 1.4.44
tables           : 3.7.0
tabulate         : 0.9.0
xarray           : 2022.11.0
xlrd             : 2.0.1
zstandard        : 0.19.0
tzdata           : None
qtpy             : None
pyqt5            : None
None

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [ ] closes #49907 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -401,7 +401,8 @@ def _wrap_applied_output(
                 not_indexed_same=not_indexed_same,
                 is_transform=is_transform,
             )
-            result.name = self.obj.name
+            if isinstance(result, Series):
```
Comment (rhshadrach) [inline pandas/core/groupby/generic.py:404]: Can you use `result.ndim == 1` here instead.

File: pandas/tests/groupby/test_apply.py [hunk 2]
```diff
@@ -334,6 +334,7 @@ def f(piece):
     result = grouped.apply(f)
 
     assert isinstance(result, DataFrame)
+    assert not hasattr(result, "name")
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_apply.py:337]: Can you add a reference to this GH issue.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Looking good!

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py:404]: sure, will do - though just for my understanding, why?

Comment (rhshadrach) [reply pandas/core/groupby/generic.py]: Interesting - it used to be that this instance checks were quite a bit more expensive than checking the `ndim` attribute; I'm now seeing isinstance is slightly cheaper. With type checkers able to naturally derive type information from isinstance checks in code paths, we should certainly prefer isinstance if they are just as performant.

Comment (rhshadrach) [reply pandas/core/groupby/generic.py:404]: @MarcoGorelli - I'm good with either using ndim or isinstance here; I plan on taking a more in depth look of our usage. Let me know what you think.

**Comment by rhshadrach** [comment]
@MarcoGorelli I added back the Bug tag. I've been doing that on regressions to differ between e.g. a performance regression and a behavior regression. Let me know if you think differently.

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (MarcoGorelli) [reply pandas/core/groupby/generic.py:404]: no preference from my end

**Comment by MarcoGorelli** [comment]
> @MarcoGorelli I added back the Bug tag. I've been doing that on regressions to differ between e.g. a performance regression and a behavior regression. Let me know if you think differently.

Hey - sounds good to me

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @MarcoGorelli!

**Pull request merged**